Index: nsec3.c
===================================================================
--- nsec3.c	(revision 3889)
+++ nsec3.c	(working copy)
@@ -820,6 +820,10 @@
 	if(!query->zone->nsec3_param)
 		return;
 	nsec3_add_nonexist_proof(query, answer, wildcard, qname);
+	if(wildcard->parent && wildcard->parent->nsec3 &&
+		wildcard->parent->nsec3->nsec3_is_exact)
+		nsec3_add_rrset(query, answer, AUTHORITY_SECTION,
+			wildcard->parent->nsec3->nsec3_cover);
 }
 
 static void
