--- linux-2.4.27/include/linux/vmalloc.h	2004-10-14 14:20:31.031658719 -0500
+++ linux/include/linux/vmalloc.h	2004-10-14 14:20:48.331203178 -0500
@@ -29,6 +29,7 @@
 extern void vmfree_area_pages(unsigned long address, unsigned long size);
 extern int vmalloc_area_pages(unsigned long address, unsigned long size,
                               int gfp_mask, pgprot_t prot);
+extern void *vcalloc(unsigned long nmemb, unsigned long elem_size);
 
 /*
  *	Allocate any pages
--- linux-2.4.27/kernel/ksyms.c	2004-10-14 14:20:31.035658152 -0500
+++ linux/kernel/ksyms.c	2004-10-14 14:20:48.336202468 -0500
@@ -115,6 +115,7 @@
 EXPORT_SYMBOL(__vmalloc);
 EXPORT_SYMBOL(vmap);
 EXPORT_SYMBOL(vmalloc_to_page);
+EXPORT_SYMBOL(vcalloc);
 EXPORT_SYMBOL(mem_map);
 EXPORT_SYMBOL(remap_page_range);
 EXPORT_SYMBOL(max_mapnr);
--- linux-2.4.27/mm/vmalloc.c	2004-10-14 14:20:31.044656874 -0500
+++ linux/mm/vmalloc.c	2004-10-14 14:20:48.340201901 -0500
@@ -382,3 +382,22 @@
 	read_unlock(&vmlist_lock);
 	return buf - buf_start;
 }
+
+void *vcalloc(unsigned long nmemb, unsigned long elem_size)
+{
+	unsigned long size;
+	void *addr;
+
+	/*
+	 * Check that we're not going to overflow.
+	 */
+	if (nmemb > (ULONG_MAX / elem_size))
+		return NULL;
+
+	size = nmemb * elem_size;
+	addr = vmalloc(size);
+	if (addr)
+		memset(addr, 0, size);
+
+	return addr;
+}
