*** gnulib/lib/javacomp.c	2006-07-11 14:08:16.000000000 +0200
--- gettext-tools/lib/javacomp.c	2006-07-30 13:39:06.000000000 +0200
***************
*** 50,56 ****
  #include "clean-temp.h"
  #include "error.h"
  #include "xvasprintf.h"
! #include "strstr.h"
  #include "gettext.h"
  
  #define _(str) gettext (str)
--- 50,56 ----
  #include "clean-temp.h"
  #include "error.h"
  #include "xvasprintf.h"
! #include "c-strstr.h"
  #include "gettext.h"
  
  #define _(str) gettext (str)
***************
*** 570,576 ****
  	  fclose (fp);
  	  goto failed;
  	}
!       envjavac_gcj = (strstr (line, "gcj") != NULL);
  
        fclose (fp);
  
--- 570,576 ----
  	  fclose (fp);
  	  goto failed;
  	}
!       envjavac_gcj = (c_strstr (line, "gcj") != NULL);
  
        fclose (fp);
  
