*** gnulib/lib/linebreak.c	2006-07-22 17:27:19.000000000 +0200
--- gettext-tools/lib/linebreak.c	2005-05-21 00:19:09.000000000 +0200
***************
*** 30,38 ****
  
  #include "utf8-ucs4.h"
  
  #include "utf16-ucs4.h"
  
- #ifdef unused
  static inline int
  u32_mbtouc (unsigned int *puc, const unsigned int *s, size_t n)
  {
--- 30,38 ----
  
  #include "utf8-ucs4.h"
  
+ #ifdef unused
  #include "utf16-ucs4.h"
  
  static inline int
  u32_mbtouc (unsigned int *puc, const unsigned int *s, size_t n)
  {
***************
*** 460,465 ****
--- 460,467 ----
  }
  
  
+ #ifdef unused
+ 
  /* Determine number of column positions required for first N units
     (or fewer if S ends before this) in S.  */
  
***************
*** 533,538 ****
--- 535,542 ----
    return width;
  }
  
+ #endif
+ 
  
  /* Determine the line break points in S, and store the result at p[0..n-1].  */
  /* We don't support line breaking of complex-context dependent characters
***************
*** 741,746 ****
--- 745,752 ----
      }
  }
  
+ #ifdef unused
+ 
  void
  u16_possible_linebreaks (const unsigned short *s, size_t n, const char *encoding, char *p)
  {
***************
*** 961,966 ****
--- 967,974 ----
      }
  }
  
+ #endif
+ 
  
  /* Choose the best line breaks, assuming the uc_width function.
     Return the column after the end of the string.  */
***************
*** 1049,1054 ****
--- 1057,1064 ----
    return last_column + piece_width;
  }
  
+ #ifdef unused
+ 
  int
  u16_width_linebreaks (const unsigned short *s, size_t n,
                        int width, int start_column, int at_end_columns,
***************
*** 1216,1221 ****
--- 1226,1233 ----
    return last_column + piece_width;
  }
  
+ #endif
+ 
  
  #ifdef TEST1
  
***************
*** 1481,1486 ****
--- 1493,1500 ----
  
  #endif /* C_CTYPE_ASCII */
  
+ #if defined unused || defined TEST2
+ 
  void
  mbs_possible_linebreaks (const char *s, size_t n, const char *encoding,
                           char *p)
***************
*** 1574,1579 ****
--- 1588,1595 ----
      }
  }
  
+ #endif
+ 
  int
  mbs_width_linebreaks (const char *s, size_t n,
                        int width, int start_column, int at_end_columns,
