*** gnulib/lib/strerror.c	2005-09-19 19:28:15.000000000 +0200
--- srclib/strerror.c	2005-10-10 13:00:57.000000000 +0200
***************
*** 21,26 ****
--- 21,28 ----
  # include <config.h>
  #endif
  
+ #if !HAVE_STRERROR
+ 
  #include <limits.h>
  
  /* Don't include <stdio.h>, since it may or may not declare
***************
*** 47,49 ****
--- 49,59 ----
    else
      return sys_errlist[n];
  }
+ 
+ #else
+ 
+ /* This declaration is solely to ensure that after preprocessing
+    this file is never empty.  */
+ typedef int dummy;
+ 
+ #endif
